src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala

```diff
@@ -0,0 +1,143 @@
+          }
+        }
+      }
+    )
+
+  protected def petGoogleIamDao(token: String)(implicit executionContext: ExecutionContext): GoogleIamDAO = {
+    new HttpGoogleIamDAO(applicationName, Token(() => token), "google")
+  }
+
+  protected def checkUserAccessFromIam(userEmail: WorkbenchEmail, userToken: OAuth2BearerToken, googleProject: GoogleProject)(implicit executionContext: ExecutionContext): Future[Boolean] = {
+    val iamDAO: GoogleIamDAO = petGoogleIamDao(userToken.token)
+    iamDAO.testIamPermission(googleProject, requiredPermissions).map {
+      foundPermissions => {
+        foundPermissions == requiredPermissions
+      }
+    }
+  }
+
+  protected def checkUserAccess(userInfo: UserInfo, googleProject: GoogleProject)(implicit executionContext: ExecutionContext): Future[Boolean] = {
+    if (cacheEnabled) {
+      notebookAuthCache.get(ProjectAuthCacheKey(userInfo.userEmail, userInfo.accessToken, googleProject, executionContext))
+    } else {
+      checkUserAccessFromIam(userInfo.userEmail, userInfo.accessToken, googleProject)
+    }
+  }
+
+  /**
+    * @param userInfo The user in question
+    * @param action The project-level action (above) the user is requesting
+    * @param googleProject The Google project to check in
+    * @return If the given user has permissions in this project to perform the specified action.
+    */
+  override def hasProjectPermission(userInfo: UserInfo, action: ProjectAction, googleProject: GoogleProject)(implicit executionContext: ExecutionContext): Future[Boolean]  = {
+    checkUserAccess(userInfo, googleProject)
+  }
+
+  /**
+    * @param userInfo The user in question
+    * @param action   The cluster-level action (above) the user is requesting
+    * @param clusterName The user-provided name of the Dataproc cluster
+    * @return If the userEmail has permission on this individual notebook cluster to perform this action
+    */
+  override def hasNotebookClusterPermission(userInfo: UserInfo, action: NotebookClusterAction, googleProject: GoogleProject, clusterName: ClusterName)(implicit executionContext: ExecutionContext): Future[Boolean]  = {
+    checkUserAccess(userInfo, googleProject)
+  }
+
+  /**
+    * Leo calls this method when it receives a "list clusters" API call, passing in all non-deleted clusters from the database.
+    * It should return a list of clusters that the user can see according to their authz.
+    *
+    * @param userInfo The user in question
+    * @param clusters All non-deleted clusters from the database
+    * @return         Filtered list of clusters that the user is allowed to see
+    */
+  override def filterUserVisibleClusters(userInfo: UserInfo, clusters: List[(GoogleProject, ClusterName)])(implicit executionContext: ExecutionContext): Future[List[(GoogleProject, ClusterName)]] = {
+    // Check each project for user-access exactly once, then filter by project.
+    val projects = clusters.map(lv => lv._1).toSet.toList
```

Comment:
Minor: the final `.toList` is probably not required

Reply:
Fixed.